src/dos/cdrom_win32.cpp

```diff
@@ -0,0 +1,229 @@
+/*
+ *  Copyright (C) 2024-2024  The DOSBox Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#if defined(WIN32)
+
+#include "cdrom.h"
+
+#include <Windows.h>
+#include <Ntddcdrm.h>
+
+static char get_drive_letter(const char* path)
+{
+	const size_t len = strlen(path);
+	if (len < 2 || len > 3) {
+		return 0;
+	}
+	if (path[1] != ':') {
+		return 0;
+	}
+	const char drive_letter = toupper(path[0]);
+	if (drive_letter >= 'A' && drive_letter <= 'Z') {
+		return drive_letter;
+	}
+	return 0;
+}
+
+bool CDROM_Interface_Win32::IsOpen() const
+{
+	return cdrom_handle != INVALID_HANDLE_VALUE;
+}
+
+CDROM_Interface_Win32::~CDROM_Interface_Win32()
+{
+	if (IsOpen()) {
+		CloseHandle(cdrom_handle);
+	}
+	cdrom_handle = INVALID_HANDLE_VALUE;
+}
+
+bool CDROM_Interface_Win32::Open(const char drive_letter)
+{
+	const std::string device_path = std::string("\\\\.\\") + drive_letter + ':';
+
+	HANDLE device = CreateFileA(device_path.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING, 0, NULL);
+	if (device == INVALID_HANDLE_VALUE) {
+		return false;
+	}
+
+	// Test to make sure this device is a CDROM drive
+	CDROM_TOC toc = {};
+	if (!DeviceIoControl(device, IOCTL_CDROM_READ_TOC, NULL, 0, &toc, sizeof(toc), NULL, NULL)) {
+		CloseHandle(device);
+		return false;
+	}
+
+	if (IsOpen()) {
+		CloseHandle(cdrom_handle);
+	}
+	cdrom_handle = device;
+	return true;
+}
+
+bool CDROM_Interface_Win32::SetDevice(const char* path)
+{
+	const char drive_letter = get_drive_letter(path);
+	if (!drive_letter) {
+		return false;
+	}
+	if (!Open(drive_letter)) {
+		return false;
+	}
+	InitAudio();
+	return true;
+}
+
+bool CDROM_Interface_Win32::GetUPC(unsigned char& attr, std::string& upc)
```

Comment:
const

Reply:
These are output parameters. 😄 But we could return an std::optional<struct> instead of using output parameters. That also ensures the caller looks at the return value.